### PR TITLE
fix: sanitize target URLs in client before using

### PR DESCRIPTION
**Pull request by adrienlaws**
# What

- checks and sanitizes URL strings before opening them from the client with a browser

# Changes

- checks for
  - empty URL string
    - returns if so
  - spaces in the URL
    - returns if so
  - lack of `HTTP/HTTPS` prefix
    - adds `HTTPS://` if there is no prefix
- adds relevant debug messages in client

### CODE REVIEW

File: clientd3d/web.c [hunk 1]
```diff
@@ -19,9 +19,34 @@ using namespace std;
...
+        debug(("URL contains spaces\n"));
+        return;
+    }
+
+    if (urlStr.find("https://") != 0 && urlStr.find("https://") != 0) {
+        urlStr = "https://" + urlStr;
+    }
+
+    HINSTANCE result = ShellExecute(NULL, "open", urlStr.c_str(), NULL, NULL, SW_SHOWNORMAL);
+    if ((INT_PTR)result <= 32) {
```
Comment (akirmse) [inline clientd3d/web.c:41]: This check doesn't do anything useful; remove.

File: clientd3d/web.h [hunk 2]
```diff
@@ -39,7 +39,8 @@ enum AccountStatusWebResponse
     FAILURE = 1 << 2
 };
 
-void WebLaunchBrowser(char *url);
+void WebLaunchBrowser(const char *url);
+bool ContainsSpaces(const std::string& str);
```
Comment (akirmse) [inline clientd3d/web.h:43]: This doesn't need to be in the header; we don't want other files calling ContainsSpaces.  Move it to the .c file and make it static.

File: clientd3d/web.c [hunk 3]
```diff
@@ -19,9 +19,34 @@ using namespace std;
...
+    }
+
+    std::string urlStr(url);
+
+    if (ContainsSpaces(urlStr)) {
+        debug(("URL contains spaces\n"));
+        return;
+    }
+
+    if (urlStr.find("https://") != 0 && urlStr.find("https://") != 0) {
```
Comment (akirmse) [inline clientd3d/web.c:36]: Doing the same check twice?  Is one of these supposed to be http?

File: clientd3d/web.c (hunk 1 above)
Comment (adrienlaws) [reply clientd3d/web.c]: Removed

File: clientd3d/web.h (hunk 2 above)
Comment (adrienlaws) [reply clientd3d/web.h]: Understood.  Made static and moved to before its WebLaunchBrowser caller.  Removed from header file.

File: clientd3d/web.c (hunk 3 above)
Comment (adrienlaws) [reply clientd3d/web.c]: Yes, mistake.  Changed to 2 separate checks (one should be HTTP, the other should be HTTPS).

File: clientd3d/web.c [hunk 4]
```diff
@@ -15,13 +15,34 @@
...
+    if (ContainsSpaces(urlStr)) {
+        debug(("URL contains spaces\n"));
+        return;
+    }
+
+    if (urlStr.find("http://") != 0 && urlStr.find("https://") != 0) {
+        urlStr = "https://" + urlStr;
+    }
+
+    HINSTANCE result = ShellExecute(NULL, "open", urlStr.c_str(), NULL, NULL, SW_SHOWNORMAL);
```
Comment (akirmse) [inline clientd3d/web.c:45]: result unused; remove

**Pull request merged**